tests/extract/test_sources.py

```diff
@@ -819,3 +819,15 @@ def some_data():
     assert list(s) == []
     info = str(s)
     assert "Source is already iterated" in info
+
+
+def test_exhausted_after_first_iter_fetch() -> None:
+
+    def some_data():
+        yield from [1, 2, 3, 4]
+
+    s = DltSource("source", "module", Schema("source"), [dlt.resource(some_data())])
```

Comment:
all good, but please add more resource types to source
- a regular list (iterable)
- an iterator (ie. on a list)
- join gen resource to transformer

I just want to make sure all cases are passing, you can do it on one test